github-stats/fetcher/client.go

```diff
@@ -0,0 +1,251 @@
+func (client *Client) limitsCheckAndWait() {
+	var sleep time.Duration
+	githubClient, err := client.getGithubClient()
+	if err != nil {
+		glog.Errorf("Failed to get Github client: %v", err)
+		sleep = time.Minute
+	} else {
+		limits, _, err := githubClient.RateLimits()
+		if err != nil {
+			glog.Errorf("Failed to get RateLimits: %v", err)
+			sleep = time.Minute
+		}
+		if limits != nil && limits.Core != nil && limits.Core.Remaining < tokenLimit {
+			sleep = limits.Core.Reset.Sub(time.Now())
+			glog.Warningf("RateLimits: reached. Sleeping for %v", sleep)
+		}
+	}
+
+	time.Sleep(sleep)
+}
+
+// ClientInterface describes what a client should be able to do
+type ClientInterface interface {
+	FetchIssues(time.Time, chan *github.Issue)
+	FetchIssueEvents(*int, chan *github.IssueEvent)
+	FetchIssueComments(int, time.Time, chan *github.IssueComment)
+	FetchPullComments(int, time.Time, chan *github.PullRequestComment)
+}
+
+// FetchIssues from Github, until 'latest' time
+func (client *Client) FetchIssues(latest time.Time, c chan *github.Issue) {
+	opt := &github.IssueListByRepoOptions{Since: latest, Sort: "updated", State: "all", Direction: "asc"}
+
+	githubClient, err := client.getGithubClient()
+	if err != nil {
+		close(c)
+		glog.Error(err)
+	}
+
+	for {
+		client.limitsCheckAndWait()
+
+		issues, resp, err := githubClient.Issues.ListByRepo(client.Org, client.Project, opt)
+		if err != nil {
+			close(c)
+			glog.Error(err)
+		}
+
+		for _, issue := range issues {
+			fmt.Println("Issue", *issue.Number, "last updated", *issue.UpdatedAt)
+			c <- issue
+		}
+
+		if resp.NextPage == 0 {
+			break
+		}
+		opt.ListOptions.Page = resp.NextPage
+	}
+
+	close(c)
+}
+
+// Look for a specific Id in a list of events
+func wasIdFound(events []*github.IssueEvent, id int) bool {
+	for _, event := range events {
+		if *event.ID == id {
+			return true
+		}
+	}
+	return false
+}
+
+// FetchIssueEvents from github and return the full list, until it matches 'latest'
+// The entire last page will be included so you can have redundancy.
+func (client *Client) FetchIssueEvents(latest *int, c chan *github.IssueEvent) {
+	opt := &github.ListOptions{PerPage: 100}
+
+	githubClient, err := client.getGithubClient()
+	if err != nil {
+		close(c)
+		glog.Error(err)
+	}
+
+	for {
+		client.limitsCheckAndWait()
+
+		glog.Infof("Downloading events page: ", opt.Page)
+		events, resp, err := githubClient.Issues.ListRepositoryEvents(client.Org, client.Project, opt)
+		if err != nil {
+			glog.Errorf("Request failed. Wait before trying again.")
+			time.Sleep(time.Second)
+			continue
+		}
+
+		for _, event := range events {
+			c <- event
+		}
+		if resp.NextPage == 0 || (latest != nil && wasIdFound(events, *latest)) {
+			break
+		}
+		opt.Page = resp.NextPage
+	}
+
+	close(c)
+}
+
+// FetchIssueComments fetches comments associated to given Issue (since latest)
+func (client *Client) FetchIssueComments(issueId int, latest time.Time, c chan *github.IssueComment) {
```

Comment:
You litteraly describing the inital code :-)

BUT:
- There are 0.4M entries
- Writing to database takes some time
- We reach token rate-limiting before fetching everything initially

Fetching all the comments from scratch takes initially about 5 hours. I don't want to hold a big (huge?) buffer and then write everything at once.